File: parser-ruby/compiler/bytecode.rb
require_relative '../utils'

class Bytecode
	BYTECODES = []

	class << self
		# Don't let others declare new bytecodes.
		private :new

		# Only this class can declare new bytecodes.
		private def declare_bytecode(opcode, name, *args, jump: false)
			raise "duplicate opcode: #{opcode}" if from_opcode opcode

			bytecode = new(name.to_s.downcase, opcode, args, jump)
			const_set name, bytecode
			BYTECODES.push bytecode
		end
	end

	def self.from_opcode(opcode)
		BYTECODES.find { _1.opcode == opcode }
	end

	# Readers for all our attributes.
	attr_reader :name, :opcode, :args

	# Iniitalize the bytecode and freeze all values.
	def initialize(name, opcode, args, is_jump)
		# Sanity check.
		raise TypeError unless name.is_a?(String) && opcode.u8? && args.is_a?(Array)

		@name, @opcode, @args, @is_jump = name.freeze, opcode, args.freeze, is_jump
		freeze
	end

	# Whether this is a jump instruction or not
	def jump?; @is_jump end

	# We should be able to use this as if it were an integer.
	alias to_i opcode
	alias to_int to_i
	def u8?; true end
	alias u8! itself

	# Our string representation's simply our name.
	alias to_s name

	# Provide a debug inspect method.
	def inspect; "#{to_s}(#@opcode, #{@args.inspect})" end

	# General bytecodes
	declare_bytecode  0, :NOP
	declare_bytecode  1, :PUSH,    :reg
	declare_bytecode  2, :PUSHI,   :i64
	declare_bytecode  3, :PUSHBSX, :i64
	declare_bytecode  4, :POP,     :reg
	declare_bytecode  5, :LOAD,    :reg, :reg
	declare_bytecode  6, :STORE,   :reg, :reg
	declare_bytecode  7, :STOREW,  :reg, :i64
	declare_bytecode  8, :STOREB,  :reg, :u8
	declare_bytecode  9, :MOV,     :reg, :reg
	declare_bytecode 10, :MOVW,    :reg, :i64
	declare_bytecode 11, :MOVBSX,  :reg, :i8
	declare_bytecode 12, :DEBUG
	declare_bytecode 13, :TRAP,    :i8,  :u8
	# Math bytecodes
	declare_bytecode 14, :INC, :reg
	declare_bytecode 15, :DEC, :reg
	declare_bytecode 16, :ADD, :reg, :reg
	declare_bytecode 17, :SUB, :reg, :reg
	declare_bytecode 18, :MUL, :reg, :reg
	declare_bytecode 19, :DIV, :reg, :reg
	declare_bytecode 20, :MOD, :reg, :reg
	declare_bytecode 21, :AND, :reg, :reg
	declare_bytecode 22, :OR,  :reg, :reg
	declare_bytecode 23, :XOR, :reg, :reg
	declare_bytecode 24, :LSH, :reg, :reg
	declare_bytecode 25, :RSH, :reg, :reg
	declare_bytecode 26, :NEG, :reg
	declare_bytecode 27, :NOT, :reg
	declare_bytecode 28, :INV, :reg

	# Jump bytecodes
	declare_bytecode 29, :CMP, :reg, :reg
	declare_bytecode 30, :JEQ, :i16, jump: true
	declare_bytecode 31, :JNE, :i16, jump: true
	declare_bytecode 32, :JLT, :i16, jump: true
	declare_bytecode 33, :JLE, :i16, jump: true
	declare_bytecode 34, :JGT, :i16, jump: true
	declare_bytecode 35, :JGE, :i16, jump: true
	declare_bytecode 36, :JMP, :i16, jump: true

	# Function bytecodes
	declare_bytecode 37, :CALLS, :i16
	declare_bytecode 38, :CALLW, :i64
	declare_bytecode 39, :RET
	declare_bytecode 40, :LFS, :reg, :u8

	declare_bytecode 255, :EXT
end

File: sjvm-old/src/bytecode.rs
use crate::{SojournVm, Flags, Word};
use std::convert::TryFrom;

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Bytecode {
	/*  General  */

	/// `Nop()`: Do nothing.
	Nop,

	/// `Push(reg)`: `push reg`.
	Push,

	/// `PushI(i64)`: `push i64`.
	PushI,

	/// `PushBSx(i8)`: `push i8`.
	PushBSx,

	/// `Pop(reg)`: `reg <- pop()`.
	Pop,

	/// `Load(dst, reg)`: `dst <- mem[reg]`.
	Load,

	/// `Store(src, reg)`: `mem[reg] <- src`.
	Store,

	/// `Mov(dst, src)`: `dst <- src`.
	Mov,

	/// `MovW(dst, i64)`: `dst <- i64`.
	MovW,

	/// `MovBSx(dst, sbyte)`: `dst <- sbyte as i64`
	MovBSx,

	/// `Debug()`: Dump the current state of the vm
	Debug,

	/// `Interrupt(what)`: Do something special depending on `what`
	Interrupt,

	/*  Math ops  */
	/// `Inc(reg)`: `reg <- reg + 1`
	Inc,

	/// `Dec(reg)`: `reg <- reg - 1`
	Dec,

	/// `Add(dst, rhs)`: `dst <- dst + rhs`.
	Add,

	/// `Sub(dst, rhs)`: `dst <- dst - rhs`.
	Sub,

	/// `Mul(dst, rhs)`: `dst <- dst * rhs`.
	Mul,

	/// `Div(dst, rhs)`: `dst <- dst / rhs`.
	Div,

	/// `Mod(dst, rhs)`: `dst <- dst % rhs`.
	Mod,

	/// `And(dst, rhs)`: `dst <- dst & rhs`.
	And,

	/// `Or(dst, rhs)`:  `dst <- dst | rhs`.
	Or,

	/// `Xor(dst, rhs)`: `dst <- dst ^ rhs`.
	Xor,

	/// `Lsh(dst, rhs)`: `dst <- dst << rhs`.
	Lsh,

	/// `Rsh(dst, rhs)`: `dst <- dst >> rhs`.
	Rsh,

	/// `Neg(dst)`: `dst <- –dst`.
	Neg,

	/// `Inv(dst)`: `dst <- ~dst`.
	Inv,

	/// `Not(dst)`: `dst <- !dst`, where `0` -> `1`, and everything else goes to zero.
	Not,

	// SAR/SOL? ROR/L?

	/*  Jumping stuff  */
	/// `Cmp(reg)`: Sets `flags` to `ZERO`, `POS`, or `NEG`, depending on what `reg` is.
	Cmp,

	/// `Jeq(i16)`: If flags are `ZERO`, then `ip <- ip + i16`.
	Jeq,

	/// `Jne(i16)`: If flags aren't `ZERO`, then  `ip <- ip + i16`.
	Jne,

	/// `Jlt(i16)`: If flags are `NEG`, then  `ip <- ip + i16`.
	Jlt,

	/// `Jle(i16)`: If flags aren't `POS`, then  `ip <- ip + i16`.
	Jle,

	/// `Jgt(i16)`: If flags are `POS`, then  `ip <- ip + i16`.
	Jgt,

	/// `Jge(i16)`: If flags aren't `NEG`, then  `ip <- ip + i16`.
	Jge,

	/// `Jmp(i16)`: `ip <- ip + i16`.
	Jmp,

	/// `Call(i16)`: `push ip ; jmp i16`
	Call,

	/// `CallF(i64)`: `push ip ; jmpw sword` (but jmpw doesn't exist lol)
	CallF,

	/// `Ret()`: `pop ip`
	Ret,

	/// `Lfs(dst, u8)`: `dst <- stack[stack.size() - 1 - u8]`.
	Lfs,

	/// Extensions that can be used later
	Extension,

}

impl Bytecode {
	/// The last Bytecode instruction.
	const LAST_BYTECODE_INSTRUCTION: Self = Self::Lfs;
}

#[derive(Debug)]
pub struct UnknownBytecode;

impl TryFrom<u8> for Bytecode {
	type Error = UnknownBytecode;

	fn try_from(byte: u8) -> Result<Self, Self::Error> {
		if byte > Self::LAST_BYTECODE_INSTRUCTION as u8 {
			Err(UnknownBytecode)
		} else {
			// SAFETY: Since all instructions are contiguous, assuming `LAST_OPCODE_INSTRUCTION` truly
			// does point to the last instruction, every `u8` here will be valid.
			unsafe {
				Ok(std::mem::transmute(byte))
			}
		}
	}
}

impl Bytecode {
	pub fn run(self, vm: &mut SojournVm) {
		// dbg!(self);
		match self {
			Self::Nop => {},
			Self::Push => vm.stack.push(vm.registers[vm.code.next_byte()]),
			Self::PushI => vm.stack.push(vm.code.next_word()),
			Self::PushBSx => vm.stack.push(vm.code.next_byte() as Word),
			Self::Pop => vm.registers[vm.code.next_byte()] = vm.stack.pop().expect("popped from empty stack!"),
			Self::Load => todo!(),
			Self::Store => todo!(),
			Self::Mov => {
				let dst = vm.code.next_byte();
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] = src;
			},
			Self::MovW => {
				let dst = vm.code.next_byte();
				vm.registers[dst] = vm.code.next_word();
			},
			Self::MovBSx => {
				let dst = vm.code.next_byte();
				vm.registers[dst] = vm.code.next_byte() as i8 as Word;
			},

			Self::Debug => { dbg!(&vm); },
			Self::Interrupt => {
				match vm.code.next_byte() {
					0 => std::process::exit(vm.code.next_short() as i32),
					1 => {
						let value = &vm.data[vm.registers[vm.code.next_byte()] as usize];
						print!("{}", String::from_utf8_lossy(value))
					},
					_ => todo!("trap")
				}
			}

			// math ops
			Self::Add => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] += src;
			}
			Self::Sub => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] -= src;
			}
			Self::Mul => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] *= src;
			}
			Self::Div => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] /= src;
			}
			Self::Mod => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] %= src;
			}
			Self::And => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] &= src;
			}
			Self::Or => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] |= src;
			}
			Self::Xor => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] ^= src;
			}
			Self::Lsh => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] <<= src;
			}
			Self::Rsh => {
				let dst = vm.code.next_byte(); 
				let src = vm.registers[vm.code.next_byte()];
				vm.registers[dst] >>= src;
			}
			Self::Neg => {
				let dst = vm.code.next_byte(); 
				vm.registers[dst] = -vm.registers[dst];
			}
			Self::Inv => {
				let dst = vm.code.next_byte(); 
				vm.registers[dst] = !vm.registers[dst];
			}
			Self::Not => {
				let dst = vm.code.next_byte(); 
				vm.registers[dst] = !(vm.registers[dst] != 0) as i64;
			},

			// jumping
			Self::Cmp => {
				vm.flags &= Flags::CMP;

				match vm.registers[vm.code.next_byte()].cmp(&0) {
					std::cmp::Ordering::Less => vm.flags |= Flags::NEG,
					std::cmp::Ordering::Equal => vm.flags |= Flags::ZERO,
					std::cmp::Ordering::Greater => vm.flags |= Flags::POS,
				}
			},

			Self::Jeq => {
				let dst = vm.code.next_short() as i16;
				if vm.flags.contains(Flags::ZERO) {
					vm.code.offset_ip(dst as isize - 3);
				}
			},
			Self::Jne => {
				let dst = vm.code.next_short() as i16;
				if !vm.flags.contains(Flags::ZERO) {
					vm.code.offset_ip(dst as isize - 3);
				}
			},
			Self::Jlt => {
				let dst = vm.code.next_short() as i16;
				if vm.flags.contains(Flags::NEG) {
					vm.code.offset_ip(dst as isize - 3);
				}
			},
			Self::Jle => {
				let dst = vm.code.next_short() as i16;
				if !vm.flags.contains(Flags::POS) {
					vm.code.offset_ip(dst as isize - 3);
				}
			},
			Self::Jgt => {
				let dst = vm.code.next_short() as i16;
				if vm.flags.contains(Flags::POS) {
					vm.code.offset_ip(dst as isize - 3);
				}
			},
			Self::Jge => {
				let dst = vm.code.next_short() as i16;
				if !vm.flags.contains(Flags::NEG) {
					vm.code.offset_ip(dst as isize - 3);
				}
			},
			Self::Jmp => {
				let dst = vm.code.next_short() as i16;
				vm.code.offset_ip(dst as isize - 3);
			},

			Self::Inc => vm.registers[vm.code.next_byte()] += 1,
			Self::Dec => vm.registers[vm.code.next_byte()] -= 1,

			Self::Call => {
				let dst = vm.code.next_sshort();
				vm.stack.push(vm.code.ip_as_word());
				vm.code.offset_ip(dst as isize - 1 - std::mem::size_of_val(&dst) as isize);
			},
			Self::CallF => {
				let dst = vm.code.next_sword();
				vm.stack.push(vm.code.ip_as_word());
				vm.code.offset_ip(dst as isize - 1 - std::mem::size_of_val(&dst) as isize);
			},
			Self::Ret => {
				let retpos = vm.stack.pop().unwrap();
				vm.code.set_ip_from_word(retpos);
			},
			Self::Lfs => {
				let dst = vm.code.next_byte();
				let which = vm.code.next_byte() as usize;
				vm.registers[dst] = vm.stack[vm.stack.len() - 2 - which]; // - 2 b/c len + ret stack ptrv
			},
			Self::Extension => todo!("extensions")
		}
	}
}

